src/events/antiSpam/messageCreate.ts

```diff
@@ -0,0 +1,170 @@
+import { Client, Constants, Message } from 'discord.js';
+import { inject, injectable } from 'tsyringe';
+import type { Event } from '../../Event';
+import { kRedis } from '../../tokens';
+import { on } from 'node:events';
+import { createHash } from 'node:crypto';
+import type { Redis } from 'ioredis';
+import { logger } from '../../logger';
+import { CaseAction, createCase } from '../../functions/cases/createCase';
+import { upsertCaseLog } from '../../functions/logs/upsertCaseLog';
+import { getGuildSetting, SettingsKeys } from '../../functions/settings/getGuildSetting';
+import i18next from 'i18next';
+import {
+	MENTION_EXPIRE_SECONDS,
+	MENTION_THRESHOLD,
+	SPAM_EXPIRE_SECONDS,
+	SPAM_SCAM_THRESHOLD,
+	SPAM_THRESHOLD,
+} from '../../Constants';
+import { checkLogChannel } from '../../functions/settings/checkLogChannel';
+
+@injectable()
+export default class implements Event {
+	public name = 'Spam check';
+
+	public event = Constants.Events.MESSAGE_CREATE;
+
+	public constructor(public readonly client: Client<true>, @inject(kRedis) public readonly redis: Redis) {}
+
+	public async execute(): Promise<void> {
+		for await (const [message] of on(this.client, this.event) as AsyncIterableIterator<[Message]>) {
+			try {
+				if (message.author.bot || !message.guild || !message.content.length) continue;
+
+				const attemptAtEveryoneOrHere = ['@everyone', '@here'].some((pattern) => message.content.includes(pattern));
+				if (message.mentions.users.size || attemptAtEveryoneOrHere) {
+					const mentionCountKey = `guild:${message.guild.id}:user:${message.author.id}:mentions`;
+
+					const increment = message.mentions.users.size + (attemptAtEveryoneOrHere ? 1 : 0);
+					let total;
+					if (await this.redis.exists(mentionCountKey)) {
+						total = await this.redis.incrby(mentionCountKey, increment);
+					} else {
+						await this.redis.setex(mentionCountKey, MENTION_EXPIRE_SECONDS, increment);
```

Comment:
There is technically a race condition here which you may want to resolve with a basic Lua script or a pipeline.

Assume 2 messages arrive at the same time:

* Message A arrives and checks if the key exists
* Message B arrives and checks if the key exists
* If the key doesn't exist, message A sets its value to `increment` with an expiration
* If the key doesn't exist, Message B sets its value to `increment` with an expiration

At the end of this scenario, the mention count is set to 1 instead of 2.

This can be solved by pipelining the `exists` call with either `incrby` or `setex` depending on the scenario. Redis will prevent other calls from happening while both calls are in-flight. Alternatively, you could write a Lua script to perform both actions (although pipelining seems to be much easier).

Reply:
good call, I think (please check) I avoided this in refactor b05afe9 with separation into util functions + less cluttered case handling.

Reply:
Yup, looks like this has been resolved.